### ISSUE TITLE
Implement "Interaction Game" Mechanism for Agents

### ISSUE THREAD

**Issue by csmangum** [Dooders/Experiments#6]
### **Description**

Add an interaction mechanism to allow agents to "play a game" with each other, such as sharing resources or attacking. This feature will enable agents to engage in complex interactions, allowing them to learn behaviors specific to their encounters with other agents.

---

### **Tasks**

1. **Expand Action Space**:
   - Add new actions for `share_resources` and `attack`.
   - Update the neural network output layer in the `DQN` class to accommodate the new actions.

2. **Enhance State Representation**:
   - Modify `get_state()` to include information about nearby agents, such as distance, angle, and resource levels.
   - Ensure state representation updates reflect relevant information for interaction decisions.

3. **Define Interaction Logic**:
   - Implement `share_resources` method:
     - Define logic for transferring a portion of resources from one agent to another.
   - Implement `attack` method:
     - Define attack logic where an agent can damage or steal resources from another agent.
   - Ensure that both actions impact resource levels and agent states appropriately.

4. **Update the `act()` Method**:
   - Modify `act()` to include new actions for interaction.
   - Add logic to select between movement and interaction actions, ensuring agents can choose to interact with nearby agents based on state information.

5. **Adjust Learning Process**:
   - Modify the reward function to account for outcomes of interactions.
     - Example: Penalize for resource loss during sharing or reward for successful attacks.
   - Ensure that reward adjustments provide meaningful feedback for the interaction actions.

6. **Implement Opponent Modeling (Optional)**:
   - Consider adding logic for agents to learn from specific agents, allowing them to adapt based on observed behaviors of nearby agents.

7. **Extend Environment to Support Interactions**:
   - Modify the environment class to support agent-to-agent interactions.
   - Define interaction range and ensure agents can detect and interact with nearby agents within this range.

8. **Testing and Debugging**:
   - Write unit tests for new methods `share_resources` and `attack`.
   - Test the overall integration of interaction actions in the simulation.
   - Add detailed logging for interaction events to track behavior changes and debug issues.

9. **Visualization and Analysis**:
   - Implement visualizations to observe interaction events and outcomes.
   - Track and display metrics like frequency of interactions, resource levels, and agent survival.

---

### **Acceptance Criteria**

- [ ] Agents have expanded action spaces, including `share_resources` and `attack`.
- [ ] Agents correctly represent nearby agent information in their state.
- [ ] Interaction logic for sharing and attacking works as expected.
- [ ] Reward function adjustments align with desired interaction outcomes.
- [ ] Environment supports agent interactions within a defined range.
- [ ] Tests cover new interaction methods and overall functionality.
- [ ] Visualizations display interaction events and resulting metrics.

---

### **Additional Notes**

- Consider exploring Multi-Agent Reinforcement Learning (MARL) algorithms if interactions lead to complex cooperation/competition dynamics.
- Ensure that the interaction mechanisms don’t introduce bias or dominance toward one strategy (e.g., attacking).
- Monitor the effect on training stability and adjust the learning process if needed.

**Comment by csmangum** [comment]
### **1. Define Interaction Actions**

First, expand the action space of your agents to include interaction actions:

- **New Actions**: Introduce actions like `share_resources`, `attack`, or `cooperate`.
- **Action Encoding**: Update the neural network's output layer to accommodate the additional actions.

**Example**:

```python
class DQN(nn.Module):
    def __init__(self, input_dim, output_dim):
        super(DQN, self).__init__()
        self.network = nn.Sequential(
            nn.Linear(input_dim, 24),
            nn.ReLU(),
            nn.Linear(24, 24),
            nn.ReLU(),
            nn.Linear(24, output_dim),  # output_dim now includes new actions
        )

    def forward(self, x):
        return self.network(x)
```

---

### **2. Update State Representation**

Modify the state to include information about nearby agents:

- **Nearby Agents**: Include data about the closest agent(s), such as their distance, relative position, and resource levels.
- **State Vector**: Extend the state vector to capture this additional information.

**Example**:

```python
def get_state(self):
    # Existing state components
    state_components = [ ... ]  # Existing state calculations

    # Get closest agent
    closest_agent = None
    min_distance = float("inf")
    for agent in self.environment.agents:
        if agent != self and agent.alive:
            dist = np.sqrt(
                (self.position[0] - agent.position[0]) ** 2
                + (self.position[1] - agent.position[1]) ** 2
            )
            if dist < min_distance:
                min_distance = dist
                closest_agent = agent

    if closest_agent:
        dx = closest_agent.position[0] - self.position[0]
        dy = closest_agent.position[1] - self.position[1]
        angle_to_agent = np.arctan2(dy, dx)
        agent_resource_level = closest_agent.resource_level / 20  # Normalized
        state_components.extend([
            min_distance / max_distance,  # Normalize distance
            angle_to_agent / np.pi,       # Normalize angle
            agent_resource_level,
        ])
    else:
        # If no agent is nearby, append zeros
        state_components.extend([0, 0, 0])

    state = torch.FloatTensor(state_components).to(self.device)
    return state
```

---

### **3. Implement Interaction Logic**

Create methods for interaction actions within your agent classes.

#### **a. Sharing Resources**

Agents can choose to share resources with nearby agents.

**Method**:

```python
def share_resources(self, target_agent):
    # Transfer resources to the target agent
    transfer_amount = min(1, self.resource_level)
    if transfer_amount > 0:
        self.resource_level -= transfer_amount
        target_agent.resource_level += transfer_amount
```

#### **b. Attacking**

Agents can attack other agents to steal resources or eliminate competition.

**Method**:

```python
def attack(self, target_agent):
    # Simple attack logic
    attack_strength = 2  # Amount of damage inflicted
    target_agent.resource_level -= attack_strength
    if target_agent.resource_level <= 0:
        target_agent.die()
```

---

### **4. Update the `act()` Method**

Modify the `act()` method to include the possibility of selecting interaction actions.

**Example**:

```python
def act(self):
    if not self.alive:
        return

    # Call parent class act() for resource consumption and death check
    super().act()
    if not self.alive:
        return

    state = self.get_state()

    # Epsilon-greedy action selection
    if random.random() < self.epsilon:
        action = random.randint(0, self.num_actions - 1)
    else:
        with torch.no_grad():
            q_values = self.model(state)
            action = q_values.argmax().item()

    self.last_state = state
    self.last_action = action

    # Map action to behavior
    if action in [0, 1, 2, 3]:
        # Movement actions (as before)
        self.move(action)
    elif action == 4:
        # Share resources
        closest_agent = self.get_closest_agent()
        if closest_agent:
            self.share_resources(closest_agent)
    elif action == 5:
        # Attack
        closest_agent = self.get_closest_agent()
        if closest_agent:
            self.attack(closest_agent)

    # Learning step will be handled after observing the outcome
```

**Note**: Update `self.num_actions` to reflect the new total number of actions (e.g., `self.num_actions = 6`).

---

### **5. Modify the Learning Process**

Ensure that the learning algorithm accounts for the new actions and their outcomes.

- **Reward Signal**: Adjust the reward function to reflect the consequences of interactions.

**Example**:

```python
def learn(self, reward):
    # ... existing code ...

    # Adjust reward based on interaction outcomes
    if self.last_action == 4:  # Share resources
        reward -= 0.5  # Sharing costs the agent resources
    elif self.last_action == 5:  # Attack
        reward += 1.0  # Reward for successful attack

    # ... existing code ...
```

**Considerations**:

- **Positive Rewards**: For successful actions (e.g., gathering resources, successful attacks).
- **Negative Rewards**: For unsuccessful actions or costs incurred (e.g., resource loss from sharing).

---

### **6. Implement Opponent Modeling (Optional)**

To enable agents to learn from interactions specific to other agents, consider implementing opponent modeling.

- **State Representation**: Include attributes of other agents (e.g., type, past behaviors).
- **Policy Adaptation**: Adjust policies based on the observed behaviors of other agents.

---

### **7. Handle Interaction Outcomes**

Ensure that the consequences of interactions are properly handled:

- **Agent Death**: If an agent's `resource_level` falls below zero due to an attack, call `die()`.
- **Resource Updates**: Update resource levels for both agents involved in the interaction.
- **State Transitions**: Capture the state transitions resulting from interactions for learning.

---

### **8. Update Environment to Support Interactions**

Modify the environment to facilitate agent interactions:

- **Agent Awareness**: Agents need access to other agents in the environment.
- **Interaction Range**: Define a maximum distance within which agents can interact.

**Example**:

```python
def get_closest_agent(self):
    max_interaction_distance = 30
    closest_agent = None
    min_distance = float("inf")
    for agent in self.environment.agents:
        if agent != self and agent.alive:
            dist = np.sqrt(
                (self.position[0] - agent.position[0]) ** 2
                + (self.position[1] - agent.position[1]) ** 2
            )
            if dist < min_distance and dist <= max_interaction_distance:
                min_distance = dist
                closest_agent = agent
    return closest_agent
```

---

### **9. Adjust Neural Network and Training**

With the increased complexity:

- **Network Architecture**: Consider expanding the network to handle the larger state and action spaces.
- **Training Stability**: Monitor training for instability due to the added complexity.

---

### **10. Testing and Debugging**

- **Unit Tests**: Write tests for new methods like `share_resources` and `attack`.
- **Simulation Runs**: Run the simulation and observe agent behaviors.
- **Logging**: Add detailed logging to track interactions and outcomes.

---

### **11. Visualization**

Visualize agent interactions to understand their behaviors:

- **Interaction Events**: Display when and where agents interact.
- **Agent Status**: Use colors or icons to indicate agent states (e.g., attacking, sharing).
- **Resource Levels**: Show resource levels to correlate actions with resource management.

---

### **12. Advanced Considerations**

#### **a. Multi-Agent Reinforcement Learning (MARL)**

- **Cooperative vs. Competitive**: Decide whether agents are learning cooperatively, competitively, or both.
- **Algorithms**: Explore MARL algorithms like MADDPG (Multi-Agent Deep Deterministic Policy Gradient) for continuous action spaces or QMIX for discrete actions.

#### **b. Communication Between Agents**

- **Signals**: Implement communication channels for agents to share information.
- **Learning to Communicate**: Allow agents to learn communication protocols.

#### **c. Social Dilemmas and Game Theory**

- **Prisoner's Dilemma**: Model interactions based on classic game theory problems.
- **Trust and Reputation**: Implement mechanisms where agents build trust or reputation scores.

---

### **13. Example Code Integration**

Here's how you might integrate the new interaction actions into your agent class:

```python
class Agent:
    def __init__(self, ...):
        # ... existing initialization ...
        self.num_actions = 6  # 4 movement actions + share + attack

    # ... existing methods ...

    def act(self):
        if not self.alive:
            return

        super().act()
        if not self.alive:
            return

        state = self.get_state()

        # Epsilon-greedy action selection
        if random.random() < self.epsilon:
            action = random.randint(0, self.num_actions - 1)
        else:
            with torch.no_grad():
                q_values = self.model(state)
                action = q_values.argmax().item()

        self.last_state = state
        self.last_action = action

        # Map action to behavior
        if action in [0, 1, 2, 3]:
            self.move(action)
        elif action == 4:
            closest_agent = self.get_closest_agent()
            if closest_agent:
                self.share_resources(closest_agent)
        elif action == 5:
            closest_agent = self.get_closest_agent()
            if closest_agent:
                self.attack(closest_agent)

    def move(self, action):
        # Implement movement logic based on action
        pass

    # ... other methods ...
```

---

### **14. Reward Function Design**

Design the reward function carefully to guide the agents toward desired behaviors.

- **Positive Rewards**:
  - Successfully gathering resources.
  - Successful attacks (if desired behavior).
  - Effective cooperation (e.g., mutual sharing leading to survival).

- **Negative Rewards**:
  - Being attacked by another agent.
  - Unsuccessful actions (e.g., attempting to share with no nearby agents).
  - Resource depletion below certain thresholds.

---

### **15. Potential Challenges**

- **Complexity**: Adding interactions increases the state and action spaces, which may require more training data and computational resources.
- **Balance**: Ensure that attacking isn't always the dominant strategy, leading to all agents attacking each other.
- **Learning Stability**: More complex environments may lead to unstable learning; consider techniques like experience replay buffers and target networks.

---

### **16. Final Thoughts**

By enabling agents to interact through sharing and attacking, you create a richer simulation environment that more closely mimics real-world scenarios. Agents can learn not only from the environment but also from the behaviors of other agents, leading to emergent strategies and complex dynamics.

**Next Steps**:

1. **Implement the Interaction Mechanisms**: Start by coding the new actions and ensuring they function as intended.
2. **Adjust the Neural Network**: Modify the network architecture and training loop to handle the new state and action spaces.
3. **Test the Simulation**: Run simulations to observe agent behaviors and make adjustments as necessary.
4. **Analyze Results**: Use logging and visualization to understand how interactions affect agent performance and learning.
5. **Iterate**: Continuously refine the agents' behaviors, reward functions, and learning algorithms based on observations.
